File: projects/client/src/networks/carbon/FormulaCompiler/parser.ts
import { assert } from '@/helpers/utilities';

import { SyntaxError } from './helpers';
import {
	AddNode,
	AnyNode,
	DivideNode,
	FunctionCallNode,
	MinusNode,
	MultiplyNode,
	NumberNode,
	PlusNode,
	StreamNode,
	SubtractNode,
} from './nodes';
import { Token, TokenType } from './tokens';

export class Parser {
	currentToken: Token | null;
	currentIndex: number;
	strIndex: number;
	lParenPos: number;
	constructor(public tokens: Token[]) {
		this.strIndex = 0;
		this.lParenPos = -1;
		this.tokens = tokens;
		this.currentIndex = -1;
		this.currentToken = null;
		this.advance();
	}

	advance() {
		this.currentIndex++;
		if (this.currentIndex < this.tokens.length) {
			this.strIndex += (this.tokens[this.currentIndex].value ?? ' ').length;
			this.currentToken = this.tokens[this.currentIndex];
		} else {
			this.currentToken = null;
		}
	}

	parse(): AnyNode | null {
		const result = this.expr();

		if (this.currentToken !== null) {
			if (this.currentToken.tokenType === TokenType.RPAREN) {
				throw new SyntaxError(`Opening parenthesis not found`, {
					startIndex: this.currentToken.startIndex,
					endIndex: this.currentToken.endIndex,
					value: this.currentToken.value,
				});
			}
			throw new SyntaxError(undefined, {
				startIndex: this.currentToken.startIndex,
				endIndex: this.currentToken.endIndex,
				value: '',
			});
		}

		return result;
	}

	expr() {
		let result: AnyNode = this.term();
		while (
			this.currentToken !== null &&
			this.currentToken?.tokenType &&
			[TokenType.PLUS, TokenType.MINUS].includes(this.currentToken.tokenType)
		) {
			if (this.currentToken.tokenType === TokenType.PLUS) {
				this.advance();
				if ([TokenType.PLUS, TokenType.MINUS].includes(this.currentToken?.tokenType)) {
					throw new SyntaxError(undefined, {
						startIndex: this.currentIndex,
						endIndex: this.currentIndex + 1,
						value: '',
					});
				}
				result = new AddNode(result, this.term());
			} else if (this.currentToken.tokenType === TokenType.MINUS) {
				this.advance();
				if ([TokenType.PLUS, TokenType.MINUS].includes(this.currentToken?.tokenType)) {
					throw new SyntaxError(undefined, {
						startIndex: this.currentIndex,
						endIndex: this.currentIndex + 1,
						value: '',
					});
				}
				result = new SubtractNode(result, this.term());
			}
		}
		return result;
	}

	term() {
		let result: AnyNode = this.factor();
		while (
			this.currentToken !== null &&
			[TokenType.MULTIPLY, TokenType.DIVIDE].includes(this.currentToken.tokenType as TokenType)
		) {
			if (this.currentToken.tokenType === TokenType.MULTIPLY) {
				this.advance();
				result = new MultiplyNode(result, this.factor());
			} else if (this.currentToken.tokenType === TokenType.DIVIDE) {
				this.advance();
				result = new DivideNode(result, this.factor());
			}
		}
		return result;
	}

	factor() {
		const token = this.currentToken;
		if (token === null)
			throw new SyntaxError(undefined, {
				startIndex: this.tokens[this.currentIndex - 1].startIndex,
				endIndex: this.tokens[this.currentIndex - 1].endIndex,
				value: '',
			});
		if (token.tokenType === TokenType.LPAREN) {
			this.advance();
			const result = this.expr();
			if (this.currentToken?.tokenType !== TokenType.RPAREN) {
				throw new SyntaxError(`Closing parenthesis not found`, {
					startIndex: token.startIndex,
					endIndex: token.endIndex,
					value: token.value,
				});
			}
			this.advance();
			return result;
		} else if (token.tokenType === TokenType.FUNCTION_HANDLE) {
			return this.functionCallFactor();
		} else if (token.tokenType === TokenType.STREAM) {
			this.advance();
			assert(token.value !== null, 'Token value is null');
			return new StreamNode(token.value);
		} else if (token.tokenType === TokenType.NUMBER) {
			this.advance();
			assert(token.value !== null, 'Token value is null');
			return new NumberNode(token.value);
		} else if (token?.tokenType === TokenType.PLUS) {
			this.advance();
			return new PlusNode(this.factor());
		} else if (token?.tokenType === TokenType.MINUS) {
			this.advance();
			return new MinusNode(this.factor());
		} else if (token?.tokenType === TokenType.RPAREN) {
			throw new SyntaxError(`Opening parenthesis not found`, {
				startIndex: token.startIndex,
				endIndex: token.endIndex,
				value: token.value,
			});
		}
		throw new SyntaxError(undefined, {
			startIndex: token.startIndex,
			endIndex: token.endIndex,
			value: token.value,
		});
	}

	functionCallFactor(): FunctionCallNode {
		const functionHandleToken = this.currentToken;
		assert(functionHandleToken !== null, 'Token value is null');
		this.advance();
		if (!this.currentToken) {
			throw new SyntaxError(undefined, {
				startIndex: this.currentIndex,
				endIndex: this.currentIndex + 1,
				value: '',
			});
		}
		if (this.currentToken.tokenType !== TokenType.LPAREN) {
			throw new SyntaxError(`Expected opening parenthesis`, {
				startIndex: this.currentToken.startIndex,
				endIndex: this.currentToken.endIndex,
				value: this.currentToken.value,
			});
		}
		this.lParenPos = this.strIndex;
		this.advance();
		const functionInputNode = this.functionInput();
		this.advance();

		//@ts-expect-error NOTE: The advance() call above will update this.currentToken, but TS doesn't know that
		if (this.currentToken?.tokenType !== TokenType.RPAREN) {
			throw new SyntaxError(`Expected closing parenthesis`, {
				startIndex: this.lParenPos - 1 ?? this.strIndex,
				endIndex: this.lParenPos ?? this.strIndex + 1,
				value: '',
			});
		}

		this.advance();
		assert(functionHandleToken.value !== null, 'Token value is null');
		return new FunctionCallNode(functionHandleToken.value, functionInputNode);
	}

	functionInput(): NumberNode {
		if (this.currentToken?.tokenType === TokenType.NUMBER) {
			assert(this.currentToken.value !== null, 'Token value is null');
			return new NumberNode(this.currentToken.value);
		}
		throw new SyntaxError(`Function input not found`, {
			startIndex: this.strIndex - (this.tokens[this.currentIndex - 1].value ?? ' ').length,
			endIndex: this.strIndex,
			value: '',
		});
	}
}

File: packages/python/combocurve/combocurve/science/network_module/parser/parser.py
from combocurve.science.network_module.parser.tokens import TokenType, Token
from combocurve.science.network_module.parser.nodes import (AllNodesTyping, NumberNode, AddNode, SubtractNode,
                                                            MultiplyNode, DivideNode, PositiveNode, NegativeNode,
                                                            StreamNode, FunctionCallNode)


class SyntaxError(Exception):
    expected = True


class Parser:
    '''
    See https://www.youtube.com/playlist?list=PLZQftyCk7_Sdu5BFaXB_jLeJ9C78si5_3 for detail

    Grammar:
    expr ::= term [ +|- term]*
    term ::= factor [ *|/ factor]*
    factor ::= ( expr ) | function_call | stream | number | + factor | - factor
    function_call ::= function_handle ( number )
    '''
    def __init__(self, tokens):
        self.tokens = tokens
        self.current_index = -1
        self.current_token: None | Token = None
        self.advance()

    def raise_error(self):
        raise SyntaxError("Invalid syntax")

    def advance(self):
        self.current_index += 1
        if self.current_index < len(self.tokens):
            self.current_token = self.tokens[self.current_index]
        else:
            self.current_token = None

    def parse(self) -> AllNodesTyping:
        '''
        Get the AST for the tokens.
        Get an expression, if there's more tokens left after the expression, then there must be syntax error
        '''
        ## try to get an expression, if there's more
        result: AllNodesTyping = self.expr()

        if self.current_token is not None:
            self.raise_error()

        return result

    def expr(self) -> AllNodesTyping:
        result: AllNodesTyping = self.term()

        while self.current_token is not None and self.current_token.token_type in (TokenType.PLUS, TokenType.MINUS):
            if self.current_token.token_type == TokenType.PLUS:
                self.advance()
                result = AddNode(result, self.term())
            elif self.current_token.token_type == TokenType.MINUS:
                self.advance()
                result = SubtractNode(result, self.term())

        return result

    def term(self) -> AllNodesTyping:
        result: AllNodesTyping = self.factor()

        while self.current_token is not None and self.current_token.token_type in (TokenType.MULTIPLY,
                                                                                   TokenType.DIVIDE):
            if self.current_token.token_type == TokenType.MULTIPLY:
                self.advance()
                result = MultiplyNode(result, self.factor())
            elif self.current_token.token_type == TokenType.DIVIDE:
                self.advance()
                result = DivideNode(result, self.factor())

        return result

    def factor(self) -> AllNodesTyping:
        token = self.current_token
        if self.current_token is None:
            self.raise_error()

        if token.token_type == TokenType.LPAREN:
            self.advance()
            result = self.expr()

            if self.current_token.token_type != TokenType.RPAREN:
                self.raise_error()

            self.advance()
            return result
        elif token.token_type == TokenType.FUNCTION_HANDLE:
            return self.function_call_factor()
        elif token.token_type == TokenType.STREAM:
            self.advance()
            return StreamNode(token.value)
        elif token.token_type == TokenType.NUMBER:
            self.advance()
            return NumberNode(token.value)

        elif token.token_type == TokenType.PLUS:
            self.advance()
            return PositiveNode(self.factor())

        elif token.token_type == TokenType.MINUS:
            self.advance()
            return NegativeNode(self.factor())
        else:
            self.raise_error()

    def function_call_factor(self) -> FunctionCallNode:
        function_handle_token = self.current_token

        ## should give left paren
        self.advance()

        if self.current_token.token_type != TokenType.LPAREN:
            self.raise_error()

        ## should give number
        self.advance()
        function_input_node = self.function_input()

        ## should give right paren
        self.advance()
        if self.current_token.token_type != TokenType.RPAREN:
            self.raise_error()

        ## advance and return
        self.advance()

        return FunctionCallNode(function_handle_token.value, function_input_node)

    def function_input(self) -> NumberNode:
        ## improve this to allow more stuff
        if self.current_token.token_type is TokenType.NUMBER:
            return NumberNode(self.current_token.value)

        else:
            self.raise_error()
